metsim/metsim.py

```diff
@@ -209,169 +226,193 @@ def _validate_force_times(self, force_times):
-            # Check that everything worked
-            [stat.get() for stat in status]
-            self.pool.join()
-            self.write(label)
+        times = self._get_output_times()
+        filename = self._get_output_filename(times)
 
-    def run(self):
+        ncout = Dataset(filename, mode="a")
+        for varname in self.params['out_vars']:
+            write_slice = (slice(None), ) + tuple(
+                self._domain_slice[d] for d in ncout.variables[varname].dimensions[1:])
+            print('writing chunk for %s-%s-%s' % (filename, varname, write_slice))
+            ncout.variables[varname][write_slice] = self.output[varname].values
+
+        ncout.sync()
+        ncout.close()
+        print('done writing chunk')
+
+    def run_slice(self):
         """
-        Kicks off the disaggregation and queues up data for IO
+        Run a single slice of
         """
         self._validate_setup()
-        iter_list = [self.met_data[dim].values
-                     for dim in self.params['iter_dims']]
         self.disagg = int(self.params['time_step']) < cnst.MIN_PER_DAY
         self.method = MetSim.methods[self.params['method']]
 
-        time_dim, groups = self._get_time_dim_and_time_dim()
-
-        for label, times in groups.items():
-            logger.info("Beginning {}".format(label))
-            self.site_generator = itertools.product(*iter_list)
-            # Add in some end point data for continuity
-            times_ext = times.union([times[0] - pd.Timedelta("1 days"),
-                                     times[-1] + pd.Timedelta("1 days")]
-                                    ).intersection(time_dim)
-            data = self.met_data.sel(time=times_ext)
-            self.setup_output(self.met_data.sel(time=times))
-
-            for site in self.site_generator:
-                locs = {k: v for k, v in zip(self.params['iter_dims'], site)}
-                # Don't run masked cells
-                if not self.domain['mask'].sel(**locs).values > 0:
-                    continue
-                wrap_results = wrap_run(
-                    self.method.run, locs, self.params, data.sel(**locs),
-                    self.state.sel(**locs), self.disagg, times, label)
-
-                # Cut the returned data down to the correct time index
-                self._unpack_results(wrap_results)
-
-            self.write(label)
-
-    def _unpack_results(self, result: tuple):
-        """Put results into the master dataset"""
-        if len(result) == 3:
-            locs, df, state = result
+        self.setup_output()
+
+        times = self.met_data['time']
+
+        params = self.params.copy()
+        for index, mask_val in np.ndenumerate(self.domain['mask'].values):
+            locs = {d: i for d, i in zip(self.domain['mask'].dims, index)}
+            print(locs)
+            # Don't run masked cells
+            if not mask_val > 0:
+                print('continuing', flush=True)
+                continue
```

Comment:
turn this into an if/else. we don't need to do anything unless mask_val is positive (i.e. don't create locs)